cmd/audiusd/main.go

```diff
@@ -3,283 +3,257 @@ package main
-		case "info":
-			slogLevel = slog.LevelInfo
-		case "warn":
-			slogLevel = slog.LevelWarn
-		case "error":
-			slogLevel = slog.LevelError
-		default:
-			slogLevel = slog.LevelWarn
-		}
-	} else {
-		slogLevel = slog.LevelInfo
-	}
+const (
+	initialBackoff = 10 * time.Second
+	maxBackoff     = 10 * time.Minute
+	maxRetries     = 10
+)
 
-	logger := common.NewLogger(&slog.HandlerOptions{
-		Level: slogLevel,
-	})
+type proxyConfig struct {
+	path   string
+	target string
+}
 
+func main() {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	sigChan := make(chan os.Signal, 1)
-	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
-
-	hostUrl, err := getHostUrl()
-	if err != nil {
-		log.Fatalf("Could not get host url: %v", err)
-	}
-
-	// make it work out of the box with no config
-	delegatePrivateKey := os.Getenv("delegatePrivateKey")
-	if delegatePrivateKey == "" {
-		delegatePrivateKey, delegateOwnerWallet := keyGen()
-		os.Setenv("delegatePrivateKey", delegatePrivateKey)
-		os.Setenv("delegateOwnerWallet", delegateOwnerWallet)
-		logger.Infof("Generated and set delegate key pair: %s", delegateOwnerWallet)
+	logger := setupLogger()
+	hostUrl := setupHostUrl()
+	setupDelegateKeyPair(logger)
+
+	services := []struct {
+		name    string
+		fn      func() error
+		enabled bool
+	}{
+		{"audiusd-echo-server", func() error { return startEchoProxyWithOptionalTLS(hostUrl, logger) }, true},
+		{"core", func() error { return core.Run(ctx, logger) }, true},
+		{"mediorum", func() error { return mediorum.Run(ctx, logger) }, isStorageEnabled()},
+		{"uptime", func() error { return uptime.Run(ctx, logger) }, hostUrl.Hostname() != "localhost"},
 	}
 
-	go func() {
-		if err := startEchoProxyWithOptionalTLS(hostUrl, tlsEnabled); err != nil {
-			log.Fatalf("Echo server failed: %v", err)
-			cancel()
-		}
-	}()
-
-	go func() {
-		if err := uptime.Run(ctx, logger); err != nil {
-			logger.Errorf("fatal uptime error: %v", err)
-			cancel()
+	for _, svc := range services {
+		if svc.enabled {
+			runWithRecover(svc.name, ctx, logger, svc.fn)
 		}
-	}()
-
-	go func() {
-		if err := core.Run(ctx, logger); err != nil {
-			logger.Errorf("fatal core error: %v", err)
-			cancel()
-		}
-	}()
-
-	if storageEnabled {
-		go func() {
-			if err := mediorum.Run(ctx, logger); err != nil {
-				logger.Errorf("fatal mediorum error: %v", err)
-				cancel()
-			}
-		}()
 	}
 
+	// Wait for shutdown signal
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
 	<-sigChan
-	logger.Info("Received termination signal, shutting down...")
 
+	logger.Info("Received termination signal, shutting down...")
 	cancel()
-
-	<-ctx.Done() // run forever, no crashloops
+	<-ctx.Done()
 	logger.Info("Shutdown complete")
 }
 
-func startEchoProxyWithOptionalTLS(hostUrl *url.URL, enableTLS bool) error {
+func runWithRecover(name string, ctx context.Context, logger *common.Logger, f func() error) {
```

Comment:
a TODO here would be implementing a way to track the status of each goroutine, such that running `audius-ctl status` could potentially tell you if mediorum/core/echo/uptime crashed